Add asynchronous Silabs USART support

Martinhoff-maker opened the pull request:
The goal of this PR is to introduce asynchronous USART in the Silabs USART driver with Silabs LDMA driver.
In order to use Silabs LDMA driver in M2P and P2M mode, some modifications need also to be made on signal binding for DMA request signal.

teburd commented:
Re-assigned to @jhedberg

jerome-pouiller commented on drivers/dma/dma_silabs_ldma.c:
```diff
@@ -485,6 +490,54 @@ static DEVICE_API(dma, dma_funcs) = {
+
+		if (LDMA_TransferDone(channel)) {
+			LDMA_StartTransfer(channel, &chan_conf->xfer_config, desc);
+		}
```
This seems racy. A bit comment would be welcome. I assume the idea is the time consumed by the CPU between `LDMA->CH[channel].LINK |= LDMA_CH_LINK_LINK` and `LDMA_TransferDone()` is always greater than the time to run the DMA.

I think it is difficult to guarantee. You have to avoid latencies on main CPU. I suggest execute the two functions inside `irq_lock()` so you won't be interrupted by IRQ (or worse, by the scheduler). However, even with that change, it is not easy to guarantee the DMA won't go faster.

Then, I worry about the atomicity of `LDMA->CH[channel].LINK`. Even if the descriptor address is already there, the CPU will re-write the whole register. Usually what happens when a peripheral and the CPU access to the same memory address in the same time is unpredictable. In my previous experiences, I saw various behavior (mix between previous value and new value, but also mix between 0 and new value).

I suggest to stress that implementation with a block of 4 bytes, a M2M DMA and a loop than add blocks indefinitely.

jerome-pouiller replied:
(BTW, I suggest to introduce `silabs_ldma_append_block()` in a separate commit)

jerome-pouiller reviewed (commented):
(I have not yet finished my review, but here is the first part)

Martinhoff-maker commented on drivers/dma/dma_silabs_ldma.c:
```diff
@@ -146,7 +149,11 @@ static int dma_silabs_block_to_descriptor(struct dma_config *config,
 	desc->xfer.doneIfs = config->complete_callback_en;
-	desc->xfer.reqMode = ldmaCtrlReqModeAll;
+	if (config->complete_callback_en) {
+		desc->xfer.reqMode = ldmaCtrlReqModeBlock;
```
@flapla After a deeper investigation, I will use "fifo_mode_control" as parameter for the request mode and btw set by default the descriptor parameter "IgnoreSRec" to 0. (I've just check this parameter and we never use it in fact so it would be better like this for now). Moreover, the description of fifo_mode_control pretty match of what ReqMode refers to. This said,concerning the SPI driver, you might have nothing to change because fifo_mode_control to 0 will select the ReqModeBlock if we are in P2M or M2P.

Martinhoff-maker committed:
test: driver: uart: add silabs slwrb4180a overlay for async usart test

Add overlay for slwrb4180a board to test async silabs usart driver with
dma.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker committed:
driver: serial: silabs: implement asynch usart with dma

First implementation of asynchronous usart driver with dma support.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker committed:
driver: dma: silabs: Add signal binding to support P2M and M2P transfer

Improve the silabs ldma driver to support P2M and M2P transfer. It also
adds signal binding to support source request binding in the dts.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker committed:
driver: dma: silabs: Add silabs_ldma_block_append function

This function allows to not restart DMA engine for driver that gives
new buffer to the DMA engine while a transfer is ongoing.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker commented:
Still making some stress test on my side concerning the silabs_ldma_append_block function. Concerning the definition of the dma slot for usart device in the .dtsi folder, I suggest to keep as it is for the moment (in an overlay file) because I fear (as nobody else do it this way) that it might introduces regression.

yishai1999 commented:
@Martinhoff-maker Were you planning on adding this to uart_silabs_eusart too? If not I will do it

Martinhoff-maker commented:
@yishai1999 yes and it's already started, PR will be open this week I hope!
